includes/validation/class-amp-site-validation.php

```diff
@@ -0,0 +1,76 @@
+<?php
+/**
+ * Class AMP_Site_Validation
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Site_Validation
+ *
+ * @since 1.0
+ */
+class AMP_Site_Validation {
+
+	/**
+	 * Gets the permalinks of all public post types with the status 'publish,' to check for AMP validity.
+	 *
+	 * This excludes 'attachment' post types, as it only searches for posts with the status 'publish.'
+	 * Attachments have a default status of 'inherit,' so they can depend on the status of their parent like a post.
+	 *
+	 * @todo: consider whether this should also return 'attachment' IDs.
+	 * @param int $number_posts The maximum amount of posts to get the permalinks for (optional).
+	 * @return string[] $permalinks The post permalinks in an array.
+	 */
+	public static function get_post_permalinks( $number_posts = 200 ) {
+		$post_types = get_post_types( array( 'public' => true ), 'names' );
+		$query      = new WP_Query( array(
+			'posts_per_page' => $number_posts,
+			'post_type'      => array_values( $post_types ),
+			'post_status'    => 'publish',
+		) );
+
+		$post_ids = wp_list_pluck( $query->posts, 'ID' );
+		return array_map( 'get_permalink', $post_ids );
+	}
+
+	/**
+	 * Gets the front-end links for public taxonomy terms, like categories and tags.
+	 *
+	 * For example, https://example.org/?cat=2
+	 * This includes categories and tags, and any more that are registered.
+	 * But it excludes post_format terms.
+	 *
+	 * @param int $number_links The maximum amount of links to get (optional).
+	 * @return string[] $links The term links in an array.
+	 */
+	public static function get_taxonomy_links( $number_links = 200 ) {
```

Comment:
There should be pagination support here. The `WP_Term_Query` supports an `$offset` arg. I suggest that this method only return links for one specific taxonomy. The site crawler will probably need to make multiple calls to such a method to iterate over all of the taxonomy terms in batches. One way to do it would be to get the terms sorted by ID in ascending order. That ensures that when you paginate through the terms you won't miss terms that get added while you're processing through them.

Reply:
Thanks, good ideas. [These commits](https://github.com/Automattic/amp-wp/compare/db54a537a5267eb49c6a80cbbdaeb1b3475ab091...26553102bc9a2f02e58b80bb9ce54159b67e6ae6) apply your suggestions, as I understand them. This now has `$taxonomy` and `$offset` parameters.